File: ports/go/core/retrieval.go
package core

import (
	"hash/fnv"
	"math"
	"sort"
	"strings"
	"unicode"
)

// Scored is a retrieval hit: id, similarity score, and the document text.
type Scored struct {
	ID    string
	Score float64
	Text  string
}

// Embed is a deterministic, model-free bag-of-words hashing embedder: tokenize, hash
// each token into one of dim buckets, count, then L2-normalize. Good enough for the
// model-free demo and identical across runs (no randomness).
func Embed(text string, dim int) []float64 {
	vec := make([]float64, dim)
	for _, tok := range tokenize(text) {
		h := fnv.New32a()
		_, _ = h.Write([]byte(tok))
		vec[h.Sum32()%uint32(dim)]++
	}
	var norm float64
	for _, v := range vec {
		norm += v * v
	}
	norm = math.Sqrt(norm)
	if norm > 0 {
		for i := range vec {
			vec[i] /= norm
		}
	}
	return vec
}

func tokenize(text string) []string {
	return strings.FieldsFunc(strings.ToLower(text), func(r rune) bool {
		return !unicode.IsLetter(r) && !unicode.IsDigit(r)
	})
}

// Cosine is the dot product of two equal-length vectors (cosine similarity when both
// are L2-normalized, as Embed produces).
func Cosine(a, b []float64) float64 {
	if len(a) != len(b) {
		return 0
	}
	var dot float64
	for i := range a {
		dot += a[i] * b[i]
	}
	return dot
}

// HotVectorIndex is the in-memory (hot tier) vector store.
type HotVectorIndex interface {
	Upsert(id string, vec []float64, text string)
	Search(query []float64, k int) []Scored
}

type hotEntry struct {
	vec  []float64
	text string
}

// InMemoryHotVectorIndex is a brute-force KNN index.
type InMemoryHotVectorIndex struct {
	entries map[string]hotEntry
}

// NewInMemoryHotVectorIndex builds an empty hot index.
func NewInMemoryHotVectorIndex() *InMemoryHotVectorIndex {
	return &InMemoryHotVectorIndex{entries: map[string]hotEntry{}}
}

func (idx *InMemoryHotVectorIndex) Upsert(id string, vec []float64, text string) {
	idx.entries[id] = hotEntry{vec: vec, text: text}
}

func (idx *InMemoryHotVectorIndex) Search(query []float64, k int) []Scored {
	hits := make([]Scored, 0, len(idx.entries))
	for id, e := range idx.entries {
		hits = append(hits, Scored{ID: id, Score: Cosine(query, e.vec), Text: e.text})
	}
	sortScored(hits)
	if k > 0 && len(hits) > k {
		hits = hits[:k]
	}
	return hits
}

// ColdSearch is the optional external (cold tier) search function.
type ColdSearch func(query []float64, k int) []Scored

// TwoTierRetriever merges hot + cold results, de-duplicating by id (hot wins) and
// returning the top-k by score.
type TwoTierRetriever struct {
	hot   HotVectorIndex
	cold  ColdSearch // may be nil
	hotK  int
	coldK int
}

// NewTwoTierRetriever builds a retriever. cold may be nil (hot-only).
func NewTwoTierRetriever(hot HotVectorIndex, cold ColdSearch, hotK, coldK int) *TwoTierRetriever {
	return &TwoTierRetriever{hot: hot, cold: cold, hotK: hotK, coldK: coldK}
}

// Retrieve returns the top-k merged hits.
func (r *TwoTierRetriever) Retrieve(query []float64, k int) []Scored {
	merged := map[string]Scored{}
	order := []string{}
	add := func(s Scored, hotTier bool) {
		if existing, ok := merged[s.ID]; ok {
			// Hot wins on text; keep the higher score.
			if hotTier {
				existing.Text = s.Text
			}
			if s.Score > existing.Score {
				existing.Score = s.Score
			}
			merged[s.ID] = existing
			return
		}
		merged[s.ID] = s
		order = append(order, s.ID)
	}
	if r.hot != nil {
		for _, s := range r.hot.Search(query, r.hotK) {
			add(s, true)
		}
	}
	if r.cold != nil {
		for _, s := range r.cold(query, r.coldK) {
			add(s, false)
		}
	}
	out := make([]Scored, 0, len(order))
	for _, id := range order {
		out = append(out, merged[id])
	}
	sortScored(out)
	if k > 0 && len(out) > k {
		out = out[:k]
	}
	return out
}

func sortScored(hits []Scored) {
	sort.SliceStable(hits, func(i, j int) bool {
		if hits[i].Score != hits[j].Score {
			return hits[i].Score > hits[j].Score
		}
		return hits[i].ID < hits[j].ID
	})
}

File: agentic-clj/src/agentic/retrieval.clj
(ns agentic.retrieval
  "Hot+cold retrieval with the deterministic FNV-1a hashing embedder — byte-compatible with the
   Python/Go/Java cores (same FNV constants, tokenization, bucketing, L2-normalization), so vectors
   and routing match across languages. Model-free."
  (:require [clojure.string :as str]))

(def ^:const fnv-offset 0x811C9DC5)
(def ^:const fnv-prime 0x01000193)
(def ^:const mask-32 0xFFFFFFFF)

(defn fnv1a-32
  "FNV-1a 32-bit hash over the UTF-8 bytes of token (returned as an unsigned long 0..2^32-1)."
  [^String token]
  (let [bytes (.getBytes token "UTF-8")
        n (alength bytes)]
    (loop [i 0 h (long fnv-offset)]
      (if (< i n)
        (let [b (bit-and (long (aget bytes i)) 0xFF)
              h (bit-and (* (bit-xor h b) fnv-prime) mask-32)]
          (recur (inc i) h))
        h))))

(defn embed
  "Bag-of-words hashing embedder, L2-normalized — a vector of dim doubles."
  [text dim]
  (let [v (double-array dim)]
    (when text
      (doseq [tok (str/split (str/lower-case text) #"[^a-z0-9]+")]
        (when (seq tok)
          (let [bucket (int (mod (fnv1a-32 tok) dim))]
            (aset v bucket (+ (aget v bucket) 1.0))))))
    (let [norm (Math/sqrt (areduce v i acc 0.0 (+ acc (* (aget v i) (aget v i)))))]
      (when (pos? norm)
        (dotimes [i dim] (aset v i (/ (aget v i) norm))))
      (vec v))))

(defn cosine [a b]
  (if (not= (count a) (count b))
    -1.0
    (let [dot (reduce + (map * a b))
          na (Math/sqrt (reduce + (map #(* % %) a)))
          nb (Math/sqrt (reduce + (map #(* % %) b)))]
      (if (or (zero? na) (zero? nb)) 0.0 (/ dot (* na nb))))))

;; ---- hot index + two-tier retriever ----

(defn hot-index [] (atom {}))

(defn upsert [idx id vec text]
  (swap! idx assoc id {:vec vec :text text}))

(defn search [idx query k]
  (->> @idx
       (map (fn [[id {:keys [vec text]}]] {:id id :score (cosine query vec) :text text}))
       (sort-by :score >)
       (take (max 1 k))
       vec))

(defn two-tier
  "A retriever merging a hot index and an optional cold search fn (cold = nil for hot-only)."
  [hot cold hot-k cold-k]
  {:hot hot :cold cold :hot-k hot-k :cold-k cold-k})

(defn retrieve [retriever query k]
  (let [hot-hits (when (:hot retriever) (search (:hot retriever) query (:hot-k retriever)))
        cold-hits (when (:cold retriever) ((:cold retriever) query (:cold-k retriever)))
        merged (->> (concat hot-hits cold-hits)
                    (group-by :id)
                    (map (fn [[_ hits]] (apply max-key :score hits)))  ; hot wins ties via max score
                    (sort-by :score >)
                    (take (max 1 k)))]
    (vec merged)))
